### PR TITLE
fixed multi-gpu inferencing issue when input size less than batch size

### PR DESCRIPTION
**Pull request by vbvg2008**

### CODE REVIEW

File: fastestimator/network.py [hunk 1]
```diff
@@ -712,10 +713,64 @@ def transform(self, data: Dict[str, Any], mode: str, epoch: int = 1) -> Dict[str
...
             data = next(iter(strategy.experimental_distribute_dataset(tf.data.Dataset.from_tensors(data))))
-        return super().transform(data, mode, epoch)
+        results = super().transform(data, mode, epoch)
+        if sub_sample:
+            results = self._subsample_data(results, batch_size)
+        return results
+
+    def _fill_batch(self, data: T, n: int) -> T:
+        """Fill data on batch dimension repeating the first n indices at the end.
+        Args:
```
Comment (TortoiseHam) [inline fastestimator/network.py:731]:
```suggestion

        Args:
```

File: fastestimator/network.py [hunk 2]
```diff
@@ -712,10 +713,64 @@ def transform(self, data: Dict[str, Any], mode: str, epoch: int = 1) -> Dict[str
...
+            data: The data to be filled.
+            The number of times to be repeated.
```
Comment (TortoiseHam) [inline fastestimator/network.py:733]:
```suggestion
            n: The number of times to be repeated.
```

File: fastestimator/network.py [hunk 3]
```diff
@@ -712,10 +713,64 @@ def transform(self, data: Dict[str, Any], mode: str, epoch: int = 1) -> Dict[str
...
+        Returns:
```
Comment (TortoiseHam) [inline fastestimator/network.py:734]:
```suggestion

        Returns:
```

File: fastestimator/network.py [hunk 4]
```diff
@@ -712,10 +713,64 @@ def transform(self, data: Dict[str, Any], mode: str, epoch: int = 1) -> Dict[str
...
+        elif isinstance(data, tuple):
+            return tuple([self._fill_batch(val, n) for val in data])
+        elif isinstance(data, set):
+            return set([self._fill_batch(val, n) for val in data])
+        elif hasattr(data, "shape"):
+            paddings = [[0, n]] + [[0, 0] for _ in range(len(data.shape) - 1)]
+            if tf.is_tensor(data):
+                return tf.pad(data, paddings=paddings, mode="symmetric")
+            else:
+                return np.pad(data, pad_width=paddings, mode="symmetric")
```
Comment (TortoiseHam) [inline fastestimator/network.py:750]: is it possible user would get torch tensors from dataset and stick them in here? or do we downcast to numpy always?

File: fastestimator/network.py [hunk 5]
```diff
@@ -712,10 +713,64 @@ def transform(self, data: Dict[str, Any], mode: str, epoch: int = 1) -> Dict[str
...
+            return data
+
+    def _subsample_data(self, data: T, n: int) -> T:
+        """Subsample data by selecting the first n indices recursively.
+        Args:
+            data: The data to be subsampled.
+            n: The number of indices to be subsampled.
+        Returns:
+            Subsampled data.
+        """
```
Comment (TortoiseHam) [inline fastestimator/network.py:761]:
```suggestion
        """Subsample data by selecting the first n indices recursively.

        Args:
            data: The data to be subsampled.
            n: The number of indices to be subsampled.

        Returns:
            Subsampled data.
        """
```

File: fastestimator/network.py (hunk 4 above)
Comment (vbvg2008) [reply fastestimator/network.py]: np.pad works on all of them.  I have removed the tf if branch

Comment (TortoiseHam) [reply fastestimator/network.py]: cause it gets turning onto a tf tensor again by the dataset?

Comment (vbvg2008) [reply fastestimator/network.py]: right.

**Review by TortoiseHam — APPROVED** [review]
LG

**Pull request merged**